src/common/url.cpp

```diff
@@ -4,19 +4,33 @@
 
 #include <common/url.h>
 
-#include <event2/http.h>
-
-#include <cstdlib>
 #include <string>
+#include <stdexcept>
+#include <charconv>
 
-std::string urlDecode(const std::string &urlEncoded) {
+std::string urlDecode(const std::string& urlEncoded) {
     std::string res;
-    if (!urlEncoded.empty()) {
-        char *decoded = evhttp_uridecode(urlEncoded.c_str(), false, nullptr);
-        if (decoded) {
-            res = std::string(decoded);
-            free(decoded);
+    res.reserve(urlEncoded.size());
+
+    for (size_t i = 0; i < urlEncoded.size(); ++i) {
+        char c = urlEncoded[i];
+        // Special handling for percent which should be followed by two hex digits
+        // representing an octet values, see RFC 3986, Section 2.1 Percent-Encoding
+        if (c == '%' && i + 2 < urlEncoded.size()) {
+            int decoded_value = 0;
+            auto [ptr, ec] = std::from_chars(&urlEncoded[i + 1], &urlEncoded[i + 3], decoded_value, 16);
+            if (ec == std::errc()) {
+                res += static_cast<char>(decoded_value);
+                // Next two characters are part of the percent encoding
+                i += 2;
+            } else {
+                // In case of invalid percent encoding, add the '%' and continue
```

Comment:
What's the rationale for continuing here? Should we accept invalid URLs? At first sight, I think I'd prefer having this return a `util::Result` and have the function fail in case of missing hex digits or invalid hex?

Reply:
I also found this a bit strange but it's the behavior of libevent and they have an explicit test case for this. I didn't investigate this further yet and I think the safest option is to keep this a pure refactor for now.

Reply:
For better or worse, this silent accepting behavior for invalid hex values in URL decoding is really common. See for example Python:

```python3
>>> import urllib.parse
>>> urllib.parse.unquote('%12%%xx%0a')
'\x12%%xx\n'
```
It's the web philopsophy of "in case of doubt, just do something!" 😅 . Anyhow, in this refactor it doesn't make sense imo to add error handling here.